Configure code quality tools

gb-beng opened the pull request.

wmelder reviewed (commented):
I cannot determine whether all the devops stuff is ok. It looks good to me, but I didn't create that part.
When Looking at the code of the server I see no fundamental changes. Mainly typing and formatting.
Using typings in Python is something that I didn't use before, so I have to get into it.
I have started the server and it returned me the right data, so that looks fine.
So, from my point of view, if @jblom and @MartijnBNG agree, this PR can be merged.

MartijnBNG commented:
From my point of view, the important part is that the data gets returned, so I'm happy. I'll leave the details of the python code to Willem and Jaap.

The updates to the github actions workflow look very good!

gb-beng commented:
Thanks for your reviews, I was still wrapping up the last bit of this branch (unify the invokation of the format/test/lint/typecheck tools by adding a script that serves as single entrypoint), hence the branch was still a draft. Will rebase, add the script and ask you for a review again to wrap up this work.

> Using typings in Python is something that I didn't use before, so I have to get into it.

@wmelder The nice thing about mypy is that you can introduce type annotations to your code incrementally... functions without them [won't be checked](https://mypy.readthedocs.io/en/stable/getting_started.html#function-signatures-and-dynamic-vs-static-typing). In that case the only new thing that might happen is warnings about imports with missing types. You can find examples in  [pyproject.toml](https://github.com/beeldengeluid/beng-lod-server/blob/config-cq-tooling/pyproject.toml) on how to ignore those (or just ask me!).

gb-beng commented:
@MartijnBNG @wmelder @jblom just converted this from draft to ready for review.

Not much changed since the earlier work you've seen (rebased, removed the new script I added and added a description to the pr) can you give a ✔️ if this looks good to you?

wmelder commented on pyproject.toml:
```diff
@@ -0,0 +1,43 @@
+  "mockito",
+  "numpy",
+  "urllib3",
+  "ontospy.*",
```
>ontospy.*",
I recently removed all the ontospy code from the repo. So this module can be removed here too. Ontospy is depending on some older packages that are causing trouble. already, We created a separate repo 'beng-lod-ontospy' to try out some solutions for the ontology visualisation.

MartijnBNG reviewed (approved):
Ziet er goed uit!

jblom commented:
@gb-beng ziet er goed uit, maar wil nog wel even iets voorzichtiger doorkijken om zeker te weten dat niet weer prod gaat chrashen, daarvoor wil ik de GH action aanpassen, zodat ie niet verder gaat als de deploy op test foutgaat

MartijnBNG commented:
> @gb-beng ziet er goed uit, maar wil nog wel even iets voorzichtiger doorkijken om zeker te weten dat niet weer prod gaat chrashen, daarvoor wil ik de GH action aanpassen, zodat ie niet verder gaat als de deploy op test foutgaat

Je kunt dat gedeelte uitcommenten, of afhankelijk maken van de vorige stap. Beide deploys hebben nu "needs: [test]" erin staan, dat kan je aanpassen.

gb-beng commented:
Converted back to draft to work on the two points above (config file + ci) and another thing I'd like to still improve in the mypy config.

wmelder commented on src/apis/mime_type_util.py:
```diff
@@ -11,20 +11,19 @@ class MimeType(Enum):
             return "json-ld"
-        elif self == MimeType.RDF_XML:
+        elif self is MimeType.RDF_XML:
             return "xml"
-        elif self == MimeType.TURTLE:
+        elif self is MimeType.TURTLE:
             return "ttl"
-        elif self == MimeType.N_TRIPLES:
+        elif self is MimeType.N_TRIPLES:
             return "nt11"
-        elif self == MimeType.N3:
+        elif self is MimeType.N3:
             return "n3"
-        elif self == MimeType.JSON:
+        elif self is MimeType.JSON:
             return "json-ld"
```
Thanks for the documentation link. Was struggling with Enums before, so it's good to get some more insight. knowing that identity instead of equality operator is recommended is helpful.

wmelder replied:
So it's still a comparison of values, but some options have the same value and are equivalent.. Can be helpful when you want to indicate that some options are actually equivalent, although they have different appearances.

wmelder replied:
we're not done yet with the enums :-{

wmelder replied:
I think it is possible to enforce unique values using a decorator: https://docs.python.org/3/library/enum.html#ensuring-unique-enumeration-values

wmelder replied:
or use auto values: https://docs.python.org/3/library/enum.html#using-automatic-values

wmelder reviewed (commented):
@gb-beng Why is the settings_bg.py file added to the repository? Do you want to make something clear? Does it mean you want changes to the settings file? Shouldn't you propose changes in the settings_example file?

wmelder commented on src/models/SDORdfConcept.py:
```diff
@@ -76,7 +78,7 @@ def __init__(self, metadata, concept_type, profile, logger, cache):
 
     # TODO: make a module for rights
-    def rights_to_license_uri(self, payload=None):
+    def rights_to_license_uri(self, payload=None):  # noqa: C901 #TODO
```
This comment is saying the function is too complex or can be written less error prone. Not necessarily wrong, but it appears it is better to refactor this a bit.

wmelder replied:
agreed to do that in another PR

wmelder commented on src/models/SDORdfConcept.py:
```diff
@@ -406,7 +408,7 @@ def __create_skos_concept(
         return skos_concept_node
 
-    def __payload_to_rdf(self, payload, parent_node, class_uri):
+    def __payload_to_rdf(self, payload, parent_node, class_uri):  # noqa: C901 #TODO
```
advice is to refactor this function/method a bit to make it less complex or error prone.

wmelder replied:
Great that you have configured it this way. It doesn't break things, but reminds us of necessary (but not urgent) improvements.

wmelder commented on src/tests/unit_tests/util/ld_util_test.py:
```diff
@@ -17,47 +23,88 @@
     [
-        (DUMMY_RESOURCE_URI, DUMMY_SPARQL_ENDPOINT, DUMMY_URI_NISV_ORGANISATION, False, True),
-        (DUMMY_RESOURCE_URI, DUMMY_SPARQL_ENDPOINT, DUMMY_URI_NISV_ORGANISATION, True, False),
+        (
```
is it necessary to put this parameter set on multiple lines? And why are not all lines split? What is our preference... it does make the tests a bit less compact and maybe less uncluttered

wmelder commented on src/util/ld_util.py:
```diff
@@ -233,12 +234,15 @@ def json_header_from_rdf_graph(
             {
                 "title": [
                     str(label)
-                    for label in rdf_graph[URIRef(resource_url) : SKOS.prefLabel]
```
Using the graphs.objects method, as is now suggested is giving the right results so this is better than the slices.

wmelder commented on src/models/SDORdfConcept.py:
```diff
@@ -592,7 +599,7 @@ def __payload_to_rdf(self, payload, parent_node, class_uri):
         return
 
-    def __parent_to_rdf(self, metadata):
+    def __parent_to_rdf(self, metadata):  # noqa: C901 #TODO
```
as with the other remarks, agreed to refactor in another PR. Let's not make this one bigger than necessary :-)

wmelder commented on src/util/ld_util.py:
```diff
@@ -85,13 +85,13 @@ def get_lod_resource_from_rdf_store(
-            publisher_present = True
-        if publisher_present is not True:
-            g.add((URIRef(resource_url), SDO.publisher, URIRef(nisv_organisation_uri)))
+        publisher_triple = (
+            URIRef(resource_url),
+            SDO.publisher,
+            URIRef(nisv_organisation_uri),
+        )
+        if publisher_triple not in g:
+            g.add(publisher_triple)
```
ok!

wmelder commented on src/apis/dataset/DataCatalogLODHandler.py:
```diff
@@ -32,7 +32,9 @@ def _init_data_catalog(self, data_catalog_file: str):
 
         repo = git.Repo(".", search_parent_directories=True)
-        git_src_dir = pathlib.Path(repo.working_tree_dir).joinpath("src")
+        git_src_dir = pathlib.Path(repo.working_tree_dir).joinpath(  # type: ignore #TODO working_tree_dir can be None
+            "src"
+        )
```
That's ok. Why are these import local? 
```python
        import git
        import pathlib

        repo = git.Repo(".", search_parent_directories=True)
        git_src_dir = pathlib.Path(repo.working_tree_dir).joinpath(  # type: ignore #TODO working_tree_dir can be None
            "src"
        )
```
By the way... I don't understand why it should be None 'if this is a bare repository'. It should still have a root dir, I would say. But I am probably not well enough informed in git's terminology.

wmelder commented on src/tests/unit_tests/util/ld_util_test.py:
```diff
@@ -23,19 +23,18 @@
+                f"{DUMMY_BENG_DATA_DOMAIN}id/{level.value}/{DUMMY_RESOURCE_ID}",
+            )
+            for level in ResourceURILevel
+        ],
```
Nice! 'Iterable unpacking', never heard of it before.

wmelder commented on src/util/ld_util.py:
```diff
@@ -176,7 +176,7 @@ def get_triples_for_blank_node_from_rdf_store(
 
     # ..and after that we merge the two together.
-    return g1 + g2
+    return g1 + g2 if (g1 and g2) else None
```
nice improvement

wmelder replied:
looks like a correction that was needed

wmelder reviewed (commented):
De query string wordt gewijzigd in een tuple met een string er in. Is dat niet een ander type en daardoor een risico dat de query in zijn geheel niet goed wordt begrepen?
```python
query_construct = (
            f"CONSTRUCT {{ ?s ?p ?o }} WHERE {{ VALUES ?s {{ <{resource_url}> }} "
            f"GRAPH <{named_graph}> {{ ?s ?p ?o FILTER(!ISBLANK(?o)) }} }}"
        )
```

wmelder reviewed (approved):
volgens mij is ie OK!
